clients/lsp-javascript.el

```diff
@@ -790,20 +790,42 @@ name (e.g. `data' variable passed as `data' parameter)."
   (when-let ((workspace (lsp-find-workspace 'ts-ls (buffer-file-name))))
     (eq 'initialized (lsp--workspace-status workspace))))
 
-(defun lsp-clients-typescript-project-ts-server-path ()
-  "Return the project local TS server path."
-  (f-join (lsp-workspace-root) "node_modules" "typescript" "lib" "tsserver.js"))
+(defun lsp-clients-typescript-require-resolve (&optional dir)
+  "Get the location of the typescript.
+Use Node.js require.
+The node_modules directory structure is suspect
+and should be trusted as little as possible.
+If you call require in Node.js,
+it should take into account the various hooks.
+For example, yarn PnP.
+
+Optional argument DIR specifies the working directory
+to run the command in."
+  (let* ((default-directory (or dir default-directory))
+         (output
+          (string-trim-right
+           (shell-command-to-string
+            "node -e \"console.log(require.resolve('typescript'))\""))))
+    (if (string-empty-p output)
+        nil
+      (f-parent output))))
+
+(defun lsp-clients-typescript-package-path-direct ()
+  "`lsp-package-path' apply `'typescript' is modified.
+because the lsp server may not recognize the tsserver executable file,
+e.g., on Windows."
+  (if (memq system-type '(cygwin windows-nt ms-dos))
+      (lsp-clients-typescript-require-resolve (f-parent (lsp-package-path 'typescript)))
+    (lsp-package-path 'typescript)))
```

Comment:
I need some clarification here. We were previously using `(f-join (f-parent (f-parent (lsp-package-path 'typescript))) "lib" "node_modules" "typescript" "lib")`, and now just `(lsp-package-path 'typescript)`? Am I missing something? Can you elaborate it? 🤔

Reply:
Until the following PR, `(lsp-package-path 'typescript)` was just called.
[lsp-javascript: supply correct path to tsserver for ts-ls by kiennq · Pull Request #4202 · emacs-lsp/lsp-mode](https://github.com/emacs-lsp/lsp-mode/pull/4202)
Maybe that was a problem in the environment of the person who made this modification, so a direct directory deep dive was done to avoid that.
I honestly don't know why different environments have different directory path locations.
My suggestion this time is to just use `(lsp-package-path 'typescript)` by default, and only use Node.js calls that return a wide range of real PATHs in special environments.
If it is fundamentally buggy, we should fix `lsp-package-path` in the first place.
By this logic, the Windows-only branching on the caller side this time is also subtle, but I was puzzled as to why the TypeScirpt server did not read the PATH with tsserver.cmd at the top level and did not know how to fix it, so I used this method as a stopgap measure. I used this method as a stopgap measure.

Reply:
Interesting thing is `(lsp-package-path 'typescript)` gives me `tsserver` and not the full path. 🤔

Reply:
I think cause by this.

```elisp
(lsp-dependency 'typescript
                '(:system "tsserver")
                '(:npm :package "typescript"
                       :path "tsserver"))
```